presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/ThriftPageSource.java

```diff
@@ -0,0 +1,194 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.connector.thrift;
+
+import com.facebook.presto.connector.thrift.api.PrestoThriftId;
+import com.facebook.presto.connector.thrift.api.PrestoThriftNullableToken;
+import com.facebook.presto.connector.thrift.api.PrestoThriftPage;
+import com.facebook.presto.connector.thrift.api.PrestoThriftService;
+import com.facebook.presto.connector.thrift.clientproviders.PrestoThriftServiceProvider;
+import com.facebook.presto.spi.ColumnHandle;
+import com.facebook.presto.spi.ConnectorPageSource;
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static io.airlift.concurrent.MoreFutures.getFutureValue;
+import static io.airlift.concurrent.MoreFutures.toCompletableFuture;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Objects.requireNonNull;
+
+public class ThriftPageSource
+        implements ConnectorPageSource
+{
+    private final PrestoThriftId splitId;
+    private final PrestoThriftService client;
+    private final List<String> columnNames;
+    private final List<Type> columnTypes;
+    private final long maxBytesPerResponse;
+    private final AtomicLong readTimeNanos = new AtomicLong(0);
+
+    private PrestoThriftId nextToken;
+    private boolean firstCall = true;
+    private CompletableFuture<PrestoThriftPage> future;
+    private long completedBytes;
+
+    public ThriftPageSource(
+            PrestoThriftServiceProvider clientProvider,
+            ThriftConnectorConfig config,
+            ThriftConnectorSplit split,
+            List<ColumnHandle> columns)
+    {
+        // init columns
+        requireNonNull(columns, "columns is null");
+        List<String> columnNames = new ArrayList<>(columns.size());
+        List<Type> columnTypes = new ArrayList<>(columns.size());
+        for (ColumnHandle columnHandle : columns) {
+            ThriftColumnHandle thriftColumnHandle = (ThriftColumnHandle) columnHandle;
+            columnNames.add(thriftColumnHandle.getColumnName());
+            columnTypes.add(thriftColumnHandle.getColumnType());
+        }
+        this.columnNames = unmodifiableList(columnNames);
+        this.columnTypes = unmodifiableList(columnTypes);
```

Comment:
I can directly specify the size of the list by doing this. Then `unmodifiableList` wrapper is cheaper than a copy with `ImmutableList`.